crates/hotshot-state-prover/src/lib.rs

```diff
@@ -1,3 +1,322 @@
+        utils::{key_pairs_for_testing, stake_table_for_testing},
+        BLSVerKey, BaseField, SchnorrVerKey, UniversalSrs,
+    };
+    use crate::{circuit::build_state_verifier_circuit, generate_state_update_proof, preprocess};
+    use ark_bn254::Bn254;
+    use ark_ec::pairing::Pairing;
+    use ark_ed_on_bn254::EdwardsConfig as Config;
+    use ark_std::{
+        rand::{CryptoRng, RngCore},
+        One,
+    };
+    use ethereum_types::U256;
+    use hotshot_stake_table::vec_based::StakeTable;
+    use hotshot_types::traits::{
+        stake_table::{SnapshotVersion, StakeTableScheme},
+        state::LightClientState,
+    };
+    use jf_plonk::{
+        errors::PlonkError,
+        proof_system::{PlonkKzgSnark, UniversalSNARK},
+        transcript::StandardTranscript,
+    };
+    use jf_primitives::{
+        crhf::{VariableLengthRescueCRHF, CRHF},
+        errors::PrimitivesError,
+        signatures::{schnorr::Signature, SchnorrSignatureScheme, SignatureScheme},
+    };
+    use jf_utils::test_rng;
+
+    // FIXME(Chengyu): see <https://github.com/EspressoSystems/jellyfish/issues/249>
+    fn universal_setup_for_testing<R>(
+        max_degree: usize,
+        rng: &mut R,
+    ) -> Result<UniversalSrs, PrimitivesError>
+    where
+        R: RngCore + CryptoRng,
+    {
+        use ark_ec::{scalar_mul::fixed_base::FixedBase, CurveGroup};
+        use ark_ff::PrimeField;
+        use ark_std::{end_timer, start_timer, UniformRand};
+
+        let setup_time = start_timer!(|| format!("KZG10::Setup with degree {}", max_degree));
+        let beta = <Bn254 as Pairing>::ScalarField::rand(rng);
+        let g = <Bn254 as Pairing>::G1::rand(rng);
+        let h = <Bn254 as Pairing>::G2::rand(rng);
+
+        let mut powers_of_beta = vec![<Bn254 as Pairing>::ScalarField::one()];
+
+        let mut cur = beta;
+        for _ in 0..max_degree {
+            powers_of_beta.push(cur);
+            cur *= &beta;
+        }
+
+        let window_size = FixedBase::get_mul_window_size(max_degree + 1);
+
+        let scalar_bits = <Bn254 as Pairing>::ScalarField::MODULUS_BIT_SIZE as usize;
+        let g_time = start_timer!(|| "Generating powers of G");
+        // TODO: parallelization
+        let g_table = FixedBase::get_window_table(scalar_bits, window_size, g);
+        let powers_of_g = FixedBase::msm::<<Bn254 as Pairing>::G1>(
+            scalar_bits,
+            window_size,
+            &g_table,
+            &powers_of_beta,
+        );
+        end_timer!(g_time);
+
+        let powers_of_g = <Bn254 as Pairing>::G1::normalize_batch(&powers_of_g);
+
+        let h = h.into_affine();
+        let beta_h = (h * beta).into_affine();
+
+        let pp = UniversalSrs {
+            powers_of_g,
+            h,
+            beta_h,
+        };
+        end_timer!(setup_time);
+        Ok(pp)
+    }
+
+    /// Internal function for helping generate the proving/verifying key
+    fn get_num_of_gates() -> Result<usize, PlonkError> {
```

Comment:
instead of this, remove it, and directly use `build_for_preprocessing()?.0.num_gates()` instead

Reply:
Done in 88784b9